packages/common/src/util.js

```diff
@@ -631,3 +633,90 @@ function getUserAgentFromPackageJson(packageJson) {
 }
 
 util.getUserAgentFromPackageJson = getUserAgentFromPackageJson;
+
+/**
+ * Wraps a callback style function to conditionally return a promise.
+ *
+ * @param {function} originalMethod - The method to promisify.
+ * @return {function} wrapped
+ */
+function promisify(originalMethod) {
+  if (originalMethod.promisified_) {
+    return originalMethod;
+  }
+
+  var slice = Array.prototype.slice;
+
+  var wrapper = function() {
+    var args = slice.call(arguments);
+    var hasCallback = is.fn(args[args.length - 1]);
+    var context = this;
+
+    if (hasCallback) {
+      return originalMethod.apply(context, args);
+    }
+
+    var PromiseCtor = PromiseOverride || Promise;
+
+    return new PromiseCtor(function(resolve, reject) {
+      args.push(function() {
+        var callbackArgs = slice.call(arguments);
+        var err = callbackArgs.shift();
+
+        if (err) {
+          return reject(err);
+        }
+
+        resolve(callbackArgs);
+      });
+
+      originalMethod.apply(context, args);
+    });
+  };
+
+  wrapper.promisified_ = true;
+  return wrapper;
+}
+
+util.promisify = promisify;
+
+/**
+ * Promisifies certain Class methods. This will not promisify private or
+ * streaming methods.
+ *
+ * @param {module:common/service} Class - Service class.
+ * @param {object=} options - Configuration object.
+ */
+function promisifyAll(Class, options) {
+  var exclude = options && options.exclude || [];
+
+  var methods = Object
+    .keys(Class.prototype)
+    .filter(function(methodName) {
+      return is.fn(Class.prototype[methodName]) && // is it a function?
+        !/(^\_|(Stream|\_)$)/.test(methodName) && // is it public/non-stream?
+        exclude.indexOf(methodName) === -1; // is it blacklisted?
+    });
+
+  methods.forEach(function(methodName) {
+    var originalMethod = Class.prototype[methodName];
+
+    if (!originalMethod.promisified_) {
+      Class.prototype[methodName] = util.promisify(originalMethod);
+    }
+  });
+}
+
+util.promisifyAll = promisifyAll;
+
+/**
+ * Allows user to override the Promise constructor without the need to touch
+ * globals. Override should be ES6 Promise compliant.
+ *
+ * @param {promise} override
+ */
+function setPromiseOverride(override) {
```

Comment:
I'm wondering if we need to localize this somehow, as a user might need to override a promise in one place and not another. Since `deduping` is default, this file is going to be re-used across all of the APIs.

And in a "ya never know" case, if a library wraps `google-cloud-node` and another library wraps *that* library, whoever sets an override would be affected all the way through.